src/misc/cloudAuth.ts

```diff
@@ -0,0 +1,199 @@
+                this.server.route({
+                    method: 'GET',
+                    path: '/callback',
+                    handler: async (request: Request) => {
+                        try {
+                            const code = request.query.code;
+                            const token = await this.fetchToken(code);
+
+                            resolve(token.access_token);
+                            return 'Thank you. You can close this tab.';
+                        } catch (err) {
+                            reject(err);
+                        } finally {
+                            this.server.stop();
+                        }
+                    },
+                });
+
+                const codeChallenge = createHash('sha256').update(this.codeVerifier).digest('base64');
+                const authorizeUrl = this.buildAuthorizeUrl(codeChallenge);
+                cli.log(chalk.green('*') + ' ' + chalk.white('...if your browser does not open, open this:')
+                    + ' ' + chalk.underline(chalk.blue(authorizeUrl)));
+
+                open(authorizeUrl);
+
+                this.server.start();
+            } catch (e) {
+                // tslint:disable-next-line:no-console
+                console.log('Error inside of the execute:', e);
+            }
+        });
+    }
+
+    public async hasToken(): Promise<boolean> {
+        await this.initialize();
+
+        return this.config.has('rcc.token.access_token');
+    }
+
+    public async getToken(): Promise<string> {
+        await this.initialize();
+
+        const item: ICloudAuthStorage = this.config.get('rcc');
+        if (new Date() < new Date(item.expiresAt)) {
+            return item.token.access_token;
+        }
+
+        await this.refreshToken();
+
+        return this.config.get('rcc.token.access_token', '') as string;
+    }
+
+    private async fetchToken(code: string | Array<string>): Promise<ICloudToken> {
+        try {
+            const request = {
+                grant_type: 'authorization_code',
+                redirect_uri: this.redirectUri,
+                client_id: clientId,
+                code,
+                code_verifier: this.codeVerifier,
+            };
+
+            const res = await axios.post(`${ cloudUrl }/api/oauth/token`, stringify(request));
+            const tokenInfo: ICloudToken = res.data;
+
+            const expiresAt = new Date();
+            expiresAt.setSeconds(expiresAt.getSeconds() + tokenInfo.expires_in);
+
+            const storageItem: ICloudAuthStorage = {
+                token: tokenInfo,
+                expiresAt,
+            };
+
+            this.config.set('rcc', storageItem);
+
+            return tokenInfo;
+        } catch (err) {
+            // tslint:disable-next-line:no-console
+            console.log(`[${ err.res.statusCode }] error getting token: ${ err.data.error } (${ err.data.requestId })`);
```

Comment:
```suggestion
            console.log(`[${ err.response.statusCode }] error getting token: ${ d.error } (${ d.requestId })`);
```

Reply:
You 100% sure about these?

Reply:
Try it.   Alter the refresh token so it’s invalid :)